spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentAlignmentSupport.scala

```diff
@@ -0,0 +1,193 @@
+    val fieldExprs = fields.zipWithIndex.map { case (field, ordinal) =>
+      Alias(GetStructField(col, ordinal, Some(field.name)), field.name)()
+    }
+
+    // recursively apply this method on nested fields
+    val newUpdates = updates.map(u => u.copy(nameParts = u.nameParts.tail))
+    val updatedFieldExprs = applyUpdates(fieldExprs, newUpdates, resolver, namePrefix :+ col.name)
+
+    // construct a new struct with updated field expressions
+    toNamedStruct(fields, updatedFieldExprs)
+  }
+
+  private def toNamedStruct(fields: Seq[StructField], fieldExprs: Seq[Expression]): Expression = {
+    val namedStructExprs = fields.zip(fieldExprs).flatMap { case (field, expr) =>
+      Seq(Literal(field.name), expr)
+    }
+    CreateNamedStruct(namedStructExprs)
+  }
+
+  private def hasExactMatch(
+      updates: Seq[ColumnUpdate],
+      col: NamedExpression,
+      resolver: Resolver): Boolean = {
+
+    updates.exists(assignment => isExactMatch(assignment, col, resolver))
+  }
+
+  private def isExactMatch(
+      update: ColumnUpdate,
+      col: NamedExpression,
+      resolver: Resolver): Boolean = {
+
+    update.nameParts match {
+      case Seq(namePart) if resolver(namePart, col.name) => true
+      case _ => false
+    }
+  }
+
+  protected def castIfNeeded(
+      expr: Expression,
+      dataType: DataType,
+      resolver: Resolver): Expression = expr match {
+    // some types cannot be casted from NullType (e.g. StructType)
+    case Literal(value, NullType) => Literal(value, dataType)
+    case _ =>
+      (expr.dataType, dataType) match {
+        // resolve structs by name if they they have the same number of fields and their names match
+        // e.g., it is ok to set a struct with fields (a, b) as another struct with fields (b, a)
+        // it is invalid to a set a struct with fields (a, d) as another struct with fields (a, b)
+        case (from: StructType, to: StructType) if requiresCast(from, to) && isValidCast(from, to, resolver) =>
+          val fieldExprs = to.flatMap { field =>
+            val fieldName = Literal(field.name)
+            val fieldExpr = ExtractValue(expr, fieldName, resolver)
+            Seq(fieldName, castIfNeeded(fieldExpr, field.dataType, resolver))
+          }
+          CreateNamedStruct(fieldExprs)
+        case (from: StructType, to: StructType) if requiresCast(from, to) =>
+          throw new AnalysisException(s"Invalid assignments: cannot cast $from to $to")
+        case (from, to) if requiresCast(from, to) => cast(expr, dataType)
+        case _ => expr
+      }
+  }
+
+  private def requiresCast(from: DataType, to: DataType): Boolean = {
+    !DataType.equalsIgnoreCaseAndNullability(from, to)
+  }
+
+  private def isValidCast(from: StructType, to: StructType, resolver: Resolver): Boolean = {
+    from.length == to.length && from.exists { f => to.exists(t => resolver(f.name, t.name))}
+  }
+
+  protected def getNameParts(expr: Expression): Seq[String] = expr match {
```

Comment:
What about `getInsertReference` or `asInsertReference`? Seems like this is actually quite specific to the references that we expect in the left side of an `INSERT` clause.

Reply:
It is also used for UPDATE SET clause too.

What about `asMultipartIdentifier`? We can make it implicit if we want to.

Reply:
Or `asAssignReference`?

Reply:
I'd go for `asAssignmentReference` and making the method implicit.